### PR TITLE
dialects: (stablehlo) add transpose

### PR DESCRIPTION
**Pull request by superlopuh**
Note stacked PR

### CODE REVIEW

File: tests/filecheck/dialects/stablehlo/ops.mlir [hunk 1]
```diff
@@ -1,17 +1,28 @@
...
+// CHECK: %multiply = "stablehlo.multiply"(%t0, %t0) : (tensor<i32>, tensor<i32>) -> tensor<i32>
+%multiply = "stablehlo.multiply"(%t0, %t0) : (tensor<i32>, tensor<i32>) -> tensor<i32>
 
-  // CHECK: %subtract = "stablehlo.subtract"(%t0, %t0) : (tensor<i32>, tensor<i32>) -> tensor<i32>
-  %subtract = "stablehlo.subtract"(%t0, %t0) : (tensor<i32>, tensor<i32>) -> tensor<i32>
-}
+// CHECK: %subtract = "stablehlo.subtract"(%t0, %t0) : (tensor<i32>, tensor<i32>) -> tensor<i32>
+%subtract = "stablehlo.subtract"(%t0, %t0) : (tensor<i32>, tensor<i32>) -> tensor<i32>
+
+%transpose_operand = "test.op"() : () -> tensor<2x3x2xi32>
```
Comment (alexarice) [inline tests/filecheck/dialects/stablehlo/ops.mlir:17]: This seems to be missing a `CHECK`?

File: xdsl/dialects/stablehlo.py [hunk 2]
```diff
@@ -126,13 +128,69 @@ class SubtractOp(ElementwiseBinaryOperation):
...
+
+    def verify_(self) -> None:
+        # Operand and result types are checked before the custom `verify_`
+        o_type = cast(TensorType[Attribute], self.operand.type)
+        r_type = cast(TensorType[Attribute], self.result.type)
+
+        o_shape = o_type.get_shape()
+        r_shape = r_type.get_shape()
+
+        # TODO: C1 - quantization
```
Comment (alexarice) [inline xdsl/dialects/stablehlo.py:169]: What do `C1` `C2` and `C3` mean here?

File: xdsl/dialects/stablehlo.py [hunk 3]
```diff
@@ -126,13 +128,69 @@ class SubtractOp(ElementwiseBinaryOperation):
...
+    name = "stablehlo.transpose"
+
+    ElementType = Annotated[Attribute, ConstraintVar("ElementType")]
+
+    operand = operand_def(TensorType[ElementType])
+    result = result_def(TensorType[ElementType])
+    permutation = attr_def(DenseArrayBase)
+
+    def __init__(
+        self, operand: SSAValue, permutation: DenseArrayBase, result_type: Attribute
```
Comment (alexarice) [inline xdsl/dialects/stablehlo.py:150]: Would it be worth also supporting inputting a permutation as a sequence of ints?

File: xdsl/dialects/stablehlo.py (hunk 2 above)
Comment (superlopuh) [reply xdsl/dialects/stablehlo.py:169]: They're the constraint names in the original stablehlo documentation

File: xdsl/dialects/stablehlo.py (hunk 3 above)
Comment (superlopuh) [reply xdsl/dialects/stablehlo.py:150]: Others may disagree, but I think that it's not a big overhead for the caller to just do this bit themselves (`DenseArrayBase.create_dense_int_or_index(i64, (2, 1, 0))`)

File: xdsl/dialects/stablehlo.py (hunk 2 above)
Comment (alexarice) [reply xdsl/dialects/stablehlo.py:169]: Would it be worth explaining this/linking in the comment?

Comment (superlopuh) [reply xdsl/dialects/stablehlo.py]: Actually I just removed them, they're not even consistently numbered throughout the document

**Review by alexarice — APPROVED** [review]
Assuming the quantization constraints can't be done because we don't have this "is_per_axis_quantized" predicate? In any case LGTM

**Comment by superlopuh** [comment]
More that I don't want to do them right now :)

**Pull request merged**